R/shared-methods.R

```diff
@@ -0,0 +1,145 @@
+#' downloading, processing, and returning data.
+dataClass <- R6::R6Class(
+  "dataClass",
+  public = list(
+    #' @field country name of country to fetch data for
+    country = "",
+    #' @field region data frame for requested region
+    region = NULL,
+    #' @field data_url link to raw data
+    data_url = "",
+    #' @field level target region level
+    level = NULL,
+    #' @field totals Boolean. If TRUE, returns totalled data per region
+    #' up to today's date.
+    totals = NULL,
+    #' @field localise Boolean. Should region names be localised.
+    localise = NULL,
+    #' @field verbose Boolean. Display information at various stages.
+    verbose = NULL,
+    #' @field steps Boolean. Keep data from each processing step.
+    steps = NULL,
+
+    #' @description General function for getting region codes for given region.
+    #' @rdname get_region_codes
+    #' @importFrom rlang .data
+    get_region_codes = function() {
+      tar_level <- paste0("level_", self$level, "_region")
+      tar_level_name <- self[[tar_level]]
+      codes <- covidregionaldata::region_codes %>%
+        filter(
+          .data$country %in% self$country,
+          .data$level %in% tar_level
+        )
+
+      if (self$verbose) {
+        message(
+          "Getting data for ", self$country,
+          " at ", tar_level_name, " administrative region"
+        )
+      }
+
+      self$region <- list(country = self$country, level = tar_level_name)
+
+      if (nrow(codes) == 1) {
+        self$region$code <- codes$name[[1]]
+        self$region$codes_lookup <- codes$codes[[1]]
+      }
+
+      self$region <- structure(
+        self$region,
+        class = c(
+          paste0("crd_", self$country, "_", self$level),
+          paste0("crd_level_", self$level)
+        )
+      )
+    },
+
+    #' @description General function for downloading raw data.
+    #' @param ... pass additional arguments
+    download_data = function(...) {
+      self$region$raw <- suppressWarnings(csv_reader(self$data_url))
+    },
+
+    #' Shared regional dataset processing
+    #'
+    #' @description General function to processes regional data.
+    #' Dynamically works for level 1 and level 2 regions.
+    #' @param ... pass additional arguments
+    process_regional = function(...) {
+      region_vars <- switch(self$level,
+        "1" = c("region_level_1", "level_1_region_code"),
+        "2" = c(
+          "region_level_2", "level_2_region_code",
+          "region_level_1", "level_1_region_code"
+        )
+      )
+      self$region <- process_regional_internal(
+        self$region,
+        group_vars = region_vars, totals = self$totals,
+        localise = self$localise, verbose = self$verbose
+      )
+    },
+
+    #' @description Optional region specific return changes.
+    #' Designed to be called after `process_regional`. For most datasets a
+    #' custom method should not be needed.
+    #' @author Sam Abbott
+    return_regional = function() {
```

Comment:
I think my expectation is that the R6 method would be returned from `get_national_data` etc along with the processed data... at the moment this is returning the old list along with assigned S3 classes (which we definitely don't want or need)

Reply:
maybe this is something for discussion as I am not really clear what the best idea is.

Reply:
#155